tests/waveplate_stub_test.py

```diff
@@ -0,0 +1,108 @@
+from pnpq.devices.waveplate_stub import WaveplateStub
+from pnpq.errors import (
+    DeviceDisconnectedError,
+    WaveplateInvalidDegreeError,
+    WaveplateInvalidMotorChannelError,
+    WaveplateInvalidStepsError,
+)
+import pytest
+
+def test_access_without_connection():
+    wp = WaveplateStub()
+    with pytest.raises(DeviceDisconnectedError):
+        wp.identify()
+
+def test_connect():
+    wp = WaveplateStub()
+    wp.connect()
+    wp.identify()
+
+def test_home():
+    wp = WaveplateStub()
+    wp.connect()
+    wp.rotate(90)
+    wp.home()
+    pos = wp.getpos()
+    assert pos == 0
+
+def test_rotate():
+    wp = WaveplateStub()
+    wp.connect()
+    wp.rotate(90)
+    pos = wp.getpos()
+    res = wp.resolution
+    assert pos == 90*res
+
+def test_rotate_invalid_degree():
+    wp = WaveplateStub()
+    wp.connect()
+    with pytest.raises(WaveplateInvalidDegreeError):
+        wp.rotate(361)
+
+def test_disable_and_enable_channels():
+    wp = WaveplateStub()
+    wp.connect()
+    wp.disable_channel(1)
+    wp.rotate(90)
+    # Should not move
+    pos = wp.getpos()
+    assert pos == 0
+    wp.enable_channel(1)
+    wp.rotate(90)
+    pos = wp.getpos()
+    res = wp.resolution
+    assert pos == 90*res
+
+def test_disable_invalid_channel():
+    wp = WaveplateStub()
+    wp.connect()
+    with pytest.raises(WaveplateInvalidMotorChannelError):
+        # Waveplates has max channel of 1
+        wp.disable_channel(2)
+
+def test_enable_duplicate_channel():
+    wp = WaveplateStub()
+    wp.connect()
+    wp.disable_channel(1)
+    wp.enable_channel(1)
+    wp.enable_channel(1)
+    # Make sure there is only one channel 1
+    assert len(wp.enabled_channels) == 1
+
+def test_step_forward_backward():
+    wp = WaveplateStub()
+    wp.connect()
+    wp.rotate(90)
+    pos = wp.getpos()
+    wp.step_forward(10)
+    newPos = wp.getpos()
+    assert newPos == pos + 10
+    wp.step_backward(10)
+    newPos = wp.getpos()
+    assert newPos == pos
+
+def test_invalid_steps():
+    wp = WaveplateStub()
+    wp.connect()
+    with pytest.raises(WaveplateInvalidStepsError):
+        wp.step_backward(1)
+        wp.step_forward(wp.max_steps + 1)
+
+def test_rotate_relative():
+    wp = WaveplateStub()
+    wp.connect()
+    wp.rotate(90)
+    pos = wp.getpos()
+    wp.rotate_relative(10)
+    newPos = wp.getpos()
+    assert newPos == pos + 10*wp.resolution
```

Comment:
The repeated use of `* wp.resolution` in these tests is a sign that the `WaveplateStub` class needs a `get_position_by_degree` method.

(Ideally `getpos` should be refactored to `get_position_by_step`).

I think this should be a follow-up GitHub task.

Reply:
I wrote a function that does exactly this in the stub (called `get_degree`), and it's used somewhat in the internal implementation. I didn't use it in the test because it's not present in the real driver (for compatibility reasons).